src/lib.rs

```diff
@@ -270,3 +298,132 @@ where
+    separator: String,
+    capacity: usize,
+}
+
+impl<'a> NamesProduct<'a, core::iter::Cycle<alloc::vec::IntoIter<Option<&'a str>>>> {
+    /// Cycles through the product of the `lists`, joining with `separator`. The
+    /// leftmost list will cycle most rapidly.
+    pub fn new(lists: &[Words<'a>], separator: &str) -> Self {
+        NamesProduct {
+            iters: lists
+                .iter()
+                .map(|words| {
+                    let mut list: Vec<Option<&'a str>> =
+                        Vec::with_capacity(words.len().saturating_add(1));
+                    list.extend(words.iter().map(|word| Some(*word)));
+                    list.push(None); // Cycle marker.
+                    (list.into_iter().cycle(), None)
+                })
+                .collect(),
+            separator: separator.to_string(),
+            capacity: Self::capacity(lists, separator),
+        }
+    }
+
+    /// Shuffles each of the given `lists` with `rng`, then cycles through the
+    /// product of the lists, joining with `separator`. The leftmost list will
+    /// cycle most rapidly.
+    pub fn shuffled<RNG>(lists: &[Words<'a>], rng: &'a mut RNG, separator: &str) -> Self
+    where
+        RNG: rand::Rng,
+    {
+        NamesProduct {
+            iters: lists
+                .iter()
+                .map(|words| {
+                    let mut list: Vec<Option<&'a str>> =
+                        Vec::with_capacity(words.len().saturating_add(1));
+                    list.extend(words.iter().map(|word| Some(*word)));
+                    list.shuffle(rng); // Could be expensive.
+                    list.push(None); // Cycle marker.
+                    (list.into_iter().cycle(), None)
+                })
+                .collect(),
+            separator: separator.to_string(),
+            capacity: Self::capacity(lists, separator),
+        }
+    }
+
+    fn capacity(lists: &[Words<'a>], separator: &str) -> usize {
+        (
+            // Sum of the length of the longest possible word in each word list.
+            lists
+                .iter()
+                .filter_map(|words| words.iter().map(|word| word.len()).max())
+                .fold(0usize, |sum, len| sum.saturating_add(len))
+            // The total length of all separators. Careful not to wrap usize.
+            + (separator.len().saturating_mul(lists.len().saturating_sub(1)))
+        )
+        // Things run _much_ quicker when the capacity is a power of 2. Memory
+        // alignment? If so it may be enough to align at, say, 8 bytes, but this
+        // works for now.
+        .checked_next_power_of_two()
+        // In case there are no lists, or they're all empty... or we have
+        // calculated that we need more than usize::MAX capacity.
+        .unwrap_or(0)
+    }
+}
+
+impl<'a, ITERATOR> Iterator for NamesProduct<'a, ITERATOR>
+where
+    ITERATOR: Iterator<Item = Option<&'a str>>,
+{
+    type Item = String;
```

Comment:
Changing this to `Rc<String>` can speed this up ~2x but the code – as I was able to implement it – was much more complex, and so I dropped it.